File: src/services/types/bot.ts

/**
 * Types for Bot Management
 */

export type BotType = 'content' | 'interaction' | 'view' | 'parser' | 'custom';

export type BotStatus = 'idle' | 'running' | 'paused' | 'error' | 'setup-required';

export type BotPlatform = 'youtube' | 'spotify' | 'instagram' | 'tiktok' | 'facebook' | 'twitter' | 'telegram' | 'multi';

export type BotHealthStatus = 'healthy' | 'warning' | 'critical' | 'unknown';

export type BotProxyStatus = 'active' | 'inactive' | 'expired' | 'unknown';

export type BotActivity = {
  id: string;
  type: 'start' | 'stop' | 'action' | 'error' | 'config';
  description: string;
  timestamp: string;
  details?: Record<string, any>;
};

export type BotConsumption = {
  cpu: number; // Percentage 0-100
  memory: number; // MB
  network: number; // KB/s
  quota: number; // Percentage of quota used 0-100
};

export type BotAction = {
  id: string;
  type: string;
  status: 'pending' | 'in-progress' | 'completed' | 'failed';
  startedAt?: string;
  completedAt?: string;
  target?: string;
  details?: Record<string, any>;
};

export type BotConfig = {
  maxActions: number;
  targetUrls?: string[];
  templates?: Record<string, string>;
  proxy?: {
    type: 'http' | 'socks' | 'none';
    url?: string;
  };
  schedule?: {
    active: boolean;
    times?: {
      start: string; // HH:MM format
      end: string; // HH:MM format
      days: number[]; // 0-6, 0 is Sunday
    }[];
  };
  advancedSettings?: Record<string, any>;
};

export type Bot = {
  id: string;
  name: string;
  type: BotType;
  platform: BotPlatform;
  status: BotStatus;
  health: BotHealthStatus;
  proxyStatus: BotProxyStatus;
  description?: string;
  avatar?: string;
  lastActive?: string;
  createdAt: string;
  updatedAt: string;
  consumption: BotConsumption;
  activity: BotActivity[];
  actions: BotAction[];
  config: BotConfig;
  associatedCampaignIds?: string[];
};

export type BotFilter = {
  type?: BotType[];
  platform?: BotPlatform[];
  status?: BotStatus[];
  health?: BotHealthStatus[];
  search?: string;
};

File: orchestrator/schemas/bot.py

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Union
from datetime import datetime
from enum import Enum

class BotStatus(str, Enum):
    IDLE = "idle"
    RUNNING = "running"
    PAUSED = "paused"
    ERROR = "error"

class BotType(str, Enum):
    VIEW = "view"
    INTERACTION = "interaction"
    COMMENT = "comment"
    SHARE = "share"

class BotPlatform(str, Enum):
    YOUTUBE = "youtube"
    INSTAGRAM = "instagram"
    TIKTOK = "tiktok"
    TWITTER = "twitter"
    SPOTIFY = "spotify"

class BotHealthStatus(str, Enum):
    HEALTHY = "healthy"
    WARNING = "warning"
    CRITICAL = "critical"
    UNKNOWN = "unknown"

class ProxyStatus(str, Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"
    ERROR = "error"

class BotBase(BaseModel):
    name: str
    type: BotType
    platform: BotPlatform
    description: Optional[str] = None
    avatar: Optional[str] = None

class BotCreate(BotBase):
    status: BotStatus = BotStatus.IDLE
    health: BotHealthStatus = BotHealthStatus.UNKNOWN
    proxy_status: ProxyStatus = ProxyStatus.INACTIVE
    config: Optional[Dict[str, Any]] = None

class BotActionBase(BaseModel):
    type: str
    status: str
    target: Optional[str] = None
    details: Optional[Dict[str, Any]] = None

class BotActionCreate(BotActionBase):
    pass

class BotAction(BotActionBase):
    id: int
    bot_id: int
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class BotActivityBase(BaseModel):
    type: str
    description: str
    details: Optional[Dict[str, Any]] = None

class BotActivityCreate(BotActivityBase):
    pass

class BotActivity(BotActivityBase):
    id: int
    bot_id: int
    timestamp: datetime

    class Config:
        from_attributes = True

class ConsumptionInfo(BaseModel):
    cpu: Optional[float] = 0
    memory: Optional[float] = 0
    network: Optional[float] = 0
    quota: Optional[float] = 0

class Bot(BotBase):
    id: int
    status: BotStatus
    health: BotHealthStatus
    proxy_status: ProxyStatus
    last_active: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    config: Optional[Dict[str, Any]] = None
    consumption: Optional[ConsumptionInfo] = None
    actions: List[BotAction] = []
    activities: List[BotActivity] = []

    class Config:
        from_attributes = True

class BotUpdate(BaseModel):
    name: Optional[str] = None
    type: Optional[BotType] = None
    platform: Optional[BotPlatform] = None
    status: Optional[BotStatus] = None
    health: Optional[BotHealthStatus] = None
    proxy_status: Optional[ProxyStatus] = None
    description: Optional[str] = None
    avatar: Optional[str] = None
    config: Optional[Dict[str, Any]] = None
